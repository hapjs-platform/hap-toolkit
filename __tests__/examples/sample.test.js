/*
 * Copyright (c) 2021-present, the hapjs-platform Project Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

const path = require('path')
const { Writable } = require('stream')
const fetch = require('node-fetch')
const fkill = require('fkill')
const stripAnsi = require('strip-ansi')
const fs = require('fs-extra')
const { run, lsfiles, readZip, wipeDynamic } = require('hap-dev-utils')
const { compile } = require('../../packages/hap-toolkit/lib')

const cwd = path.resolve(__dirname, '../../examples/sample')

describe('hap-toolkit', () => {
  const distDir = path.resolve(cwd, 'dist')
  fs.removeSync(distDir)

  it(
    'hap-build: 默认流式打包，包内存在META-INF文件',
    async () => {
      await run('npm', ['run', 'build'], [], { cwd })
      const rpks = await lsfiles('dist/*.rpk', { cwd })
      // TODO more details
      expect(rpks.length).toBe(1)
      let rpkPath = path.resolve(cwd, rpks[0])
      // 读取压缩包中的内容
      const packages = await readZip(rpkPath)
      const hasMeta = packages.files['META-INF/']
      const hasMetaCert = packages.files['META-INF/CERT']
      expect(hasMeta).toBeTruthy()
      expect(hasMetaCert).toBeTruthy()
    },
    6 * 60 * 1000
  )

  it(
    'hap-build [--disable-stream-pack]: 包内不存在META-INF文件',
    async () => {
      await run('npm', ['run', 'build', '-- --disable-stream-pack'], [], { cwd })
      const rpks = await lsfiles('dist/*.rpk', { cwd })
      let rpkPath = path.resolve(cwd, rpks[0])
      // 读取压缩包中的内容
      const packages = await readZip(rpkPath)
      const hasMeta = packages.files['META-INF/']
      expect(hasMeta).toBeFalsy()
    },
    6 * 60 * 1000
  )

  it(
    'hap-server',
    async () => {
      const dialogs = [
        {
          pattern: (output) => {
            return output.match(/生成HTTP服务器的二维码: (http:\/\/.*)/)
          },
          feeds: (proc, output) => {
            const match = output.match(/生成HTTP服务器的二维码: (http:\/\/.*)/)
            const url = match[1]
            const p1 = fetch(url)
              .then((res) => {
                expect(res.status).toBe(200)
                return res.text()
              })
              .then((text) => {
                expect(text).toMatch('<title>调试器</title>')
              })
            const p2 = fetch(url + '/qrcode')
              .then((res) => res.arrayBuffer())
              .then((buffer) => {
                expect(Buffer.from(buffer).readUInt32BE(0)).toBe(0x89504e47)
              })

            Promise.all([p1, p2]).then(async () => {
              await fkill(proc.pid)
            })
          }
        },
        {
          pattern: /listen EADDRINUSE: address already in use/,
          type: 'stderr',
          feeds: (proc, output) => {
            proc.kill('SIGINT')
            proc.kill('SIGTERM')
            throw new Error('address in use')
          }
        }
      ]

      await run('npm', ['run', 'server'], dialogs, { cwd })
    },
    6 * 60 * 1000
  )

  it(
    'missed release pem files',
    async () => {
      let happened = false
      const dialogs = [
        {
          pattern: /编译错误，缺少release签名私钥文件/,
          type: 'stderr',
          feeds: (proc, output) => {
            happened = true
          }
        }
      ]
      await run('npm', ['run', 'release'], dialogs, { cwd })
      /**
      TODO fix:
      在windows平台上， function talkTo 里面不能正确拿到完整的 stdout, stderr
       */

      expect(happened).toBe(true)
    },
    6 * 60 * 1000
  )

  // 这里会记录很多内容到 snapshots
  it(
    'compile native',
    async () => {
      const outputs = []
      const outputStream = new Writable({
        write(chunk, encoding, next) {
          outputs.push(chunk.toString())
          next()
        }
      })
      // 第三个参数为是否开启watch，true为开启
      // TODO other than `native`?
      const data = await compile('native', 'dev', false, {
        cwd,
        log: outputStream,
        buildNameFormat: 'CUSTOM=dev'
      })
      // 更详细的 snapshots
      const json = data.stats.toJson({
        source: true
      })
      const wipe = (content) =>
        wipeDynamic(content, [
          [cwd, '<project-root>'],
          [/大小为 \d+ KB/g, '大小为 <SIZE> KB']
        ])

      const sourceArr = []
      const identifierArr = []
      json.modules.forEach((module) => {
        if (module.source) {
          sourceArr.push(wipe(module.source))
          identifierArr.push(wipe(module.identifier))
        }
      })
      expect(sourceArr.join('\n\n')).toMatchSnapshot(identifierArr.join('\n\n'))
      const rpks = await lsfiles('dist/*.rpk', { cwd })
      const hasCustom = rpks.some((item) => item.indexOf('dev') !== -1)
      expect(hasCustom).toBeTruthy()
      expect(json.assets.map((a) => a.name).sort()).toMatchSnapshot('assets list')
      const output = stripAnsi(outputs.join('\n'))
      // TODO expect(wipe(output)).toMatchSnapshot('outputs')
      expect(wipe(output)).not.toBeNull()
    },
    6 * 60 * 1000
  )
})
