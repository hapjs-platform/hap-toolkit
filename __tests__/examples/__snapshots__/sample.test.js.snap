// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hap-toolkit compile native: assets list 1`] = `
"length: 96
app.js
component/basic/a/index.js
component/basic/image/index.js
component/basic/progress/index.js
component/basic/rating/index.js
component/basic/span/index.js
component/basic/text/index.js
component/container/div/index.js
component/container/list/index.js
component/container/popup/index.js
component/container/refresh/index.js
component/container/richtext/index.js
component/container/stack/index.js
component/container/swiper/index.js
component/container/tabs/index.js
component/form/input/index.js
component/form/label/index.js
component/form/multi-picker/index.js
component/form/picker/index.js
component/form/slider/index.js
component/form/switch/index.js
component/form/textarea/index.js
component/media/audio/index.js
component/media/canvas/index.js
component/media/video/index.js
component/media/web/detail/index.js
component/media/web/index.js
component/style/animation/index.js
component/style/background9image/index.js
component/style/backgroundposition/index.js
component/style/font-family/index.js
component/style/transform/index.js
component/style/translatepercent/index.js
component/thirdParty/map/index.js
framework/deeplink/index.js
framework/exit/index.js
framework/exit/pages/page1/index.js
framework/exit/pages/page2/index.js
framework/exit/pages/page3/index.js
framework/lifecycle/index.js
functionality/animationapi/index.js
functionality/ceiling/index.js
functionality/changestyle/index.js
functionality/customWeb/index.js
functionality/form/index.js
functionality/h5download/index.js
functionality/listcombination/index.js
functionality/menutop/index.js
functionality/photo/index.js
functionality/postMessage/index.js
functionality/qrcode/index.js
functionality/simulate-swipe/index.js
functionality/template/elec-business/index.js
functionality/titlebar/index.js
functionality/verification/index.js
functionality/web/index.js
home/index.js
interface/system/app/index.js
interface/system/battery/index.js
interface/system/brightness/index.js
interface/system/calendar/index.js
interface/system/clipboard/index.js
interface/system/contact/index.js
interface/system/device/index.js
interface/system/fetch/index.js
interface/system/file/index.js
interface/system/filerw/index.js
interface/system/forbid/index.js
interface/system/geolocation/index.js
interface/system/image/index.js
interface/system/media/index.js
interface/system/network/index.js
interface/system/notification/index.js
interface/system/package/index.js
interface/system/prompt/index.js
interface/system/qrcode/index.js
interface/system/record/index.js
interface/system/request/index.js
interface/system/router/detail/index.js
interface/system/router/index.js
interface/system/sensor/index.js
interface/system/share/index.js
interface/system/shortcut/index.js
interface/system/sms/index.js
interface/system/storage/index.js
interface/system/vibrator/index.js
interface/system/volume/index.js
interface/system/websocket/index.js
interface/system/webview/index.js
interface/system/wifi/index.js
interface/thirdparty/alipay/index.js
interface/thirdparty/qqaccount/index.js
interface/thirdparty/serviceshare/index.js
interface/thirdparty/wbaccount/index.js
interface/thirdparty/wxaccount/index.js
interface/thirdparty/wxpay/index.js"
`;
